app/code/Magento/Quote/Model/QuoteManagement.php

```diff
@@ -298,22 +298,39 @@ public function assignCustomer($cartId, $customerId, $storeId)
             );
         }
         try {
-            $this->quoteRepository->getForCustomer($customerId);
-            throw new StateException(
-                __("The customer can't be assigned to the cart because the customer already has an active cart.")
-            );
-        // phpcs:ignore Magento2.CodeAnalysis.EmptyBlock
+            $customerActiveQuote = $this->quoteRepository->getForCustomer($customerId);
         } catch (\Magento\Framework\Exception\NoSuchEntityException $e) {
+            /** This exception appear when customer have no active cart*/
+            $customerActiveQuote = false;
         }
 
+        if ($customerActiveQuote) {
```

Comment:
I think you had miscommunication with Stanislav. Stanislav mean that you can move this
```
 $quote->merge($customerActiveQuote);
 $customerActiveQuote->setIsActive(0);
 $this->quoteRepository->save($customerActiveQuote);
```
code block to try/catch block where you reveice `$customerActiveQuote`. In this case you will not need to have login in catch construction ( `$customerActiveQuote = false` ). 

Logic based on try/catch is bad approach.

Reply:
Thank you for clarification. Code updated, please let me know if it still require changes.